lib/internal/modules/esm/ipc_types.js

```diff
@@ -0,0 +1,99 @@
+/* eslint-disable */
+let nextId = 1;
+function getNewId() {
+  const id = nextId;
+  nextId++;
+  return id;
+}
+
+class LoaderWorker {
+  resolveImportURL() {
+    throw new TypeError('Not implemented');
+  }
+
+  getFormat() {
+    throw new TypeError('Not implemented');
+  }
+
+  getSource() {
+    throw new TypeError('Not implemented');
+  }
+
+  transformSource() {
+    throw new TypeError('Not implemented');
+  }
+
+  addBelowPort() {
+    throw new TypeError('Not implemented');
+  }
+}
+
+class RemoteLoaderWorker {
+  constructor(port) {
+    this._port = port;
+    this._pending = new Map();
+
+    this._port.on('message', this._handleMessage.bind(this));
+    this._port.unref();
+  }
+
+  _handleMessage({ id, error, result }) {
+    if (!this._pending.has(id)) return;
+    const { resolve, reject } = this._pending.get(id);
+    this._pending.delete(id);
+    this._port.unref();
```

Comment:
```suggestion
    if (!this._pending.size) {
      this._port.unref();
    }
```

The failure shown in my demo repository is caused by the following execution sequence:
```js
this._send(...); // calls this._port.ref()
this._send(...); // calls this._port.ref()
this._handleMessage(...); // calls this._port.unref()
```

The suggested change ensures we only call `this._port.unref()` when the pending queue is cleared.

Reply:
Thanks, makes sense!